src/solver/reduce/forward_sum.cpp

```diff
@@ -0,0 +1,334 @@
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+    }
+
+    {
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(output_numel, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel = KernelInfo{};
+
+        kernel.kernel_file = "MIOpenSum.cpp";
+        kernel.kernel_name = "SumFwdContiguous";
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+            {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+            {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        };
+
+        kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+    }
+
+    if(!is_num_work_item_enough && is_parallelism_enough)
+    {
+        result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+            return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+                decltype(auto) parallel_kernel = handle_.Run(kernels[0]);
+                decltype(auto) kernel          = handle_.Run(kernels[1]);
+                decltype(auto) params          = raw_params.CastTo<miopen::reduce::InvokeParams>();
+
+                auto xdims = params.xDesc->GetLengths();
+                auto ydims = params.yDesc->GetLengths();
+                auto dim   = params.dim;
+
+                auto reduce_size = xdims[dim];
+                auto output_numel =
+                    std::accumulate(ydims.begin(), ydims.end(), 1ULL, std::multiplies<size_t>());
+
+                auto inner_size = 1ULL;
+                for(int32_t i = dim + 1; i < xdims.size(); i++)
+                {
+                    inner_size *= xdims[i];
+                }
+
+                // At least 4 WGs per one CU
+                auto reqd_work_item_cnt =
+                    static_cast<size_t>(256 * handle_.GetMaxComputeUnits() * 4);
+
+                auto parallelism_size = 1ULL;
+                while(parallelism_size * output_numel < reqd_work_item_cnt &&
+                      parallelism_size < std::sqrt(reduce_size))
+                {
+                    parallelism_size *= 2ULL;
+                }
```

Comment:
It seems that computation can be moved outside of the function and reused in `GetWorkspaceSize` and in this function in another `if` branch.

Reply:
fixed it